From 0ea5e1aca0c7aab7d878933c7219207b0af77b27 Mon Sep 17 00:00:00 2001
From: Sathishkumar Muruganandam <quic_murugana@quicinc.com>
Date: Tue, 28 Nov 2023 20:06:08 +0530
Subject: [PATCH] wifi: ath12k: fix tx power, max reg power update to firmware

Currently values for max_regpower, max_power, max_antenna_gain are
multiplied by 2 before sending to firmware which is wrong since uses
1dBm step for power calculations.

This leads to wrong power values used in firmware and radio.

Fix the update of max_regpower, max_power, max_antenna_gain values
and also add missing max tx power update in SCAN CHAN LIST WMI cmd.

Signed-off-by: Sathishkumar Muruganandam <quic_murugana@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 12 ++++++------
 drivers/net/wireless/ath/ath12k/wmi.c |  2 ++
 2 files changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 206b719c6c50..0c83f436f9e8 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -11719,9 +11719,9 @@ static int ath12k_mac_vdev_start_restart(struct ath12k_link_vif *arvif,
 		ath12k_phymodes[chandef->chan->band][chandef->width];
 
 	arg.channel.min_power = 0;
-	arg.channel.max_power = chandef->chan->max_power * 2;
-	arg.channel.max_reg_power = chandef->chan->max_reg_power * 2;
-	arg.channel.max_antenna_gain = chandef->chan->max_antenna_gain * 2;
+	arg.channel.max_power = chandef->chan->max_power;
+	arg.channel.max_reg_power = chandef->chan->max_reg_power;
+	arg.channel.max_antenna_gain = chandef->chan->max_antenna_gain;
 	arg.ru_punct_bitmap = ~chandef->ru_punct_bitmap;
 
 	arg.pref_tx_streams = ar->num_tx_chains;
@@ -12161,9 +12161,9 @@ ath12k_mac_multi_vdev_restart(struct ath12k *ar,
 		ath12k_phymodes[chandef->chan->band][chandef->width];
 
 	arg.channel.min_power = 0;
-	arg.channel.max_power = chandef->chan->max_power * 2;
-	arg.channel.max_reg_power = chandef->chan->max_reg_power * 2;
-	arg.channel.max_antenna_gain = chandef->chan->max_antenna_gain * 2;
+	arg.channel.max_power = chandef->chan->max_power;
+	arg.channel.max_reg_power = chandef->chan->max_reg_power;
+	arg.channel.max_antenna_gain = chandef->chan->max_antenna_gain;
 	arg.channel.chan_radar = !!(chandef->chan->flags & IEEE80211_CHAN_RADAR);
 	arg.channel.passive = arg.channel.chan_radar;
 	arg.channel.freq2_radar = radar_enabled;
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index ac6abe95ce04..8f408e386fba 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -3918,6 +3918,8 @@ int ath12k_wmi_send_scan_chan_list_cmd(struct ath12k *ar,
 						  WMI_CHAN_REG_INFO1_REG_CLS);
 			*reg2 |= le32_encode_bits(channel_arg->antennamax,
 						  WMI_CHAN_REG_INFO2_ANT_MAX);
+			*reg2 |= le32_encode_bits(channel_arg->maxregpower,
+						  WMI_CHAN_REG_INFO2_MAX_TX_PWR);
 
 			ath12k_dbg(ar->ab, ATH12K_DBG_WMI,
 				   "WMI chan scan list chan[%d] = %u, chan_info->info %8x\n",
-- 
2.25.1

